### DISCUSSION TITLE
Cask "onyx" -- hash mismatch because wrong file is being downloaded on macos Ventura

### DISCUSSION THREAD

**Discussion by jimscard** [Homebrew/discussions#4157] [Casks] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 3.6.20-146-g4b3d64d
ORIGIN: https://github.com/Homebrew/brew
HEAD: 4b3d64d307c8aa7f364c3f854a0b4dc29d59ecd0
Last commit: 2 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 369d38a2dd04607af4199bd3cc87ef5db86fb881
Core tap last commit: 3 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.8pQL0HBKMC/org.xquartz:0
HOMEBREW_EDITOR: /usr/local/bin/codium
HOMEBREW_INSTALL_FROM_API: set
HOMEBREW_MAKE_JOBS: 16
HOMEBREW_UPDATE_REPORT_ALL_FORMULAE: set
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 14.0.0 build 1400
Git: 2.37.1 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.86.0 => /usr/bin/curl
macOS: 13.2-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: 14.2
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

When I try to upgrade the "onyx" cask, I receive a SHA 256 mismatch.  After some debugging attempts, it turns out that the reason that the mismatch is occurring is that for some reason, instead of downloading the Onyx.dmg file from the "13" directory on the server, the URL that is being used is the "12" one.  

The same thing happens if I try to reinstall … 

> brew upgrade onyx --verbose --debug
/usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/3.6.20-146-g4b3d64d\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 13.2\)\ curl/7.86.0 --header Accept-Language:\ en --fail --max-time 5 --retry 3 --location --remote-time --output /Users/jimscard/Library/Caches/Homebrew/api/formula.json --time-cond /Users/jimscard/Library/Caches/Homebrew/api/formula.json --compressed --silent https://formulae.brew.sh/api/formula.json
/usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/3.6.20-146-g4b3d64d\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 13.2\)\ curl/7.86.0 --header Accept-Language:\ en --max-time 5 --retry 3 --fail https://formulae.brew.sh/api/cask-source/onyx.rb
/usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/3.6.20-146-g4b3d64d\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 13.2\)\ curl/7.86.0 --header Accept-Language:\ en --fail --max-time 5 --retry 3 --location --remote-time --output /Users/jimscard/Library/Caches/Homebrew/api/cask.json --time-cond /Users/jimscard/Library/Caches/Homebrew/api/cask.json --compressed --silent https://formulae.brew.sh/api/cask.json
==> Upgrading 1 outdated package:
onyx 4.3.5 -> 4.3.6
==> Started upgrade process for Cask onyx
==> Upgrading onyx
==> Printing caveats
==> Cask::Installer#fetch
/usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/3.6.20-146-g4b3d64d\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 13.2\)\ curl/7.86.0 --header Accept-Language:\ en --retry 3 --location --silent --head --request GET https://www.titanium-software.fr/download/12/OnyX.dmg
==> Downloading https://www.titanium-software.fr/download/12/OnyX.dmg
/usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/3.6.20-146-g4b3d64d\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 13.2\)\ curl/7.86.0 --header Accept-Language:\ en --retry 3 --location --head https://www.titanium-software.fr/download/12/OnyX.dmg
/usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/3.6.20-146-g4b3d64d\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 13.2\)\ curl/7.86.0 --header Accept-Language:\ en --fail --retry 3 --location --remote-time --output /Users/jimscard/Library/Caches/Homebrew/downloads/750069d93a61822d8d38e45c15105350c6573e218297078827c5f45f92dd72c8--OnyX.dmg.incomplete https://www.titanium-software.fr/download/12/OnyX.dmg
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 5433k  100 5433k    0     0  31.4M      0 --:--:-- --:--:-- --:--:-- 32.7M
==> Checking quarantine support
/usr/bin/xattr -h
/usr/bin/swift -target x86_64-apple-macosx13 /usr/local/Homebrew/Library/Homebrew/cask/utils/quarantine.swift
==> Quarantine is available.
==> Verifying Gatekeeper status of /Users/jimscard/Library/Caches/Homebrew/downloads/750069d93a61822d8d38e45c15105350c6573e218297078827c5f45f92dd72c8--OnyX.dmg
/usr/bin/xattr -p com.apple.quarantine /Users/jimscard/Library/Caches/Homebrew/downloads/750069d93a61822d8d38e45c15105350c6573e218297078827c5f45f92dd72c8--OnyX.dmg
==> /Users/jimscard/Library/Caches/Homebrew/downloads/750069d93a61822d8d38e45c15105350c6573e218297078827c5f45f92dd72c8--OnyX.dmg is not quarantined
==> Quarantining /Users/jimscard/Library/Caches/Homebrew/downloads/750069d93a61822d8d38e45c15105350c6573e218297078827c5f45f92dd72c8--OnyX.dmg
/usr/bin/swift -target x86_64-apple-macosx13 /usr/local/Homebrew/Library/Homebrew/cask/utils/quarantine.swift /Users/jimscard/Library/Caches/Homebrew/downloads/750069d93a61822d8d38e45c15105350c6573e218297078827c5f45f92dd72c8--OnyX.dmg https://www.titanium-software.fr/download/12/OnyX.dmg https://www.titanium-software.fr/en/onyx.html
==> Verifying checksum for cask 'onyx'
==> Purging files for version 4.3.6 of Cask onyx
Error: onyx: SHA256 mismatch
Expected: dcf443b5585d85f8e5c03124a37a982c0885ff8826e33de93c6ffa60221bebc5
  Actual: 5d4bd7ca16fbd2bd70a383f74e6c254d281776ae0a012d1484c392cc4dc5ee28
    File: /Users/jimscard/Library/Caches/Homebrew/downloads/750069d93a61822d8d38e45c15105350c6573e218297078827c5f45f92dd72c8--OnyX.dmg
To retry an incomplete download, remove the file above.
/usr/local/Homebrew/Library/Homebrew/cask/cmd/upgrade.rb:169:in `upgrade_casks'
/usr/local/Homebrew/Library/Homebrew/cmd/upgrade.rb:226:in `upgrade_outdated_casks'
/usr/local/Homebrew/Library/Homebrew/cmd/upgrade.rb:117:in `upgrade'
/usr/local/Homebrew/Library/Homebrew/brew.rb:105:in `<main>'

### Relevant casks

onyx

**Comment by p-linnane** [comment]
This appears to be a bug with `HOMEBREW_INSTALL_FROM_API`. I'm unable to reproduce it when that isn't set.

CC @Rylan12

**Reply by juju35** [reply]
still does not work on unset this variable

**Reply by gromgit** [reply]
`export HOMEBREW_NO_INSTALL_FROM_API=1` and try again.

**Reply by juju35** [reply]
It works fine thanks :-)

**Comment by Rylan12** [comment, marked as answer]
This should now be fixed for real and should no longer need the `HOMEBREW_NO_INSTALL_FROM_API` variable to be set. Sorry for the inconvenience!
